### PR TITLE
Fix Regression on ee9 / ee8 MultiPart parsing

### PR DESCRIPTION
**Pull request by joakime**
In Jetty 10 / Jetty 11 when a `multipart/form-data` is sent with an empty request body, it would result in an error status 400 (Bad Request).

The implementation in ee9 / ee8 in Jetty 12 returns an error status 500.

Introduce test cases and fix to restore this behavior.

### CODE REVIEW

File: jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/MultiPartFormInputStream.java [hunk 1]
```diff
@@ -617,6 +618,14 @@ protected void parse()
                 }
                 else if (len == -1)
                 {
+                    if (total == 0)
+                    {
+                        throw new BadMessageException(
+                            "No progress made on multipart/form-data",
```
Comment (gregw) [inline jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/MultiPartFormInputStream.java]: This error message does not seam right.   "No progress" would be if "len == 0".  This is Early EOF or similar.

**Review by gregw — CHANGES_REQUESTED** [review]
This seams a very specific check/failure.
What if the body is a single byte? The EOF can come at many illegal points and why do we need/have special handling just for 0 bytes?

File: jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/MultiPartFormInputStream.java (hunk 1 above)
Comment (joakime) [reply jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/MultiPartFormInputStream.java]: The message is a bit awkward, yes.

Putting this a different way, the point where this case should occur is a EOF with no start of a `multipart/form-data` parse.

It could be any of the following scenarios where `Content-Type: multipart/form-data; boundary=---babcxyz` is declared.

1. an empty request body (aka `Content-Length: 0`)
2. a request body which exists (content length is non-zero) but is all whitespace

Both of those are 400 Bad Requests (in the Jetty 10 / Jetty 11 codebase)

This shouldn't catch scenarios like ...

1. A request body that is a content length smaller than the boundary length
2. A request body that opens a boundary then EOFs
3. A request body that is malformed for `multipart/form-data`

etc..
Those have traditionally been 500 error statuses.

Comment (joakime) [reply jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/MultiPartFormInputStream.java]: Thoughts on a different approach?

Comment (gregw) [reply jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/MultiPartFormInputStream.java]: @joakime you are misunderstanding me.      

As you say, there are a whole bunch of ways an multipart body can be invalid by being truncated early.    These should all result in a 400 (you said 500, which I guess is kind of understandable if some random exception is thrown, but technically we should try to turn all of them into a 400).

So what I'm asking, is why do we need a specific check for zero bytes received, if we are already handling less bytes than a boundary, EOF after boundary etc. etc.
Do we have tests for all those scenarios?   Surely we are already testing somewhere that we got a valid multi-part and that an empty body is not valid?

**Comment by joakime** [comment]
I've added more tests for this regression, and fixed many of the 500 vs 400 status returns in the process. (more are returning 400 now)

File: jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/MultiPartFormInputStream.java (hunk 1 above)
Comment (gregw) [reply jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/MultiPartFormInputStream.java]: @joakime this question is still not answered.   Why is there a specific test for `total==0`, but not one for `total==1`?     How is the later handled? and why can't that handling also handle the former?

**Review by gregw — CHANGES_REQUESTED** [review]
I think we need to hangout on this as I don't think I'm communicating my concern.

File: jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/MultiPartFormInputStream.java (hunk 1 above)
Comment (janbartel) [reply jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/MultiPartFormInputStream.java]: Are we sure this is a regression? In eg jetty-10 we had a fairly extensive suite of tests for the `MultiPartInputStreamParser` and the `MultiPartFormInputStream` classes, which I assume were ported to jetty-12?

Comment (joakime) [reply jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/MultiPartFormInputStream.java]: @janbartel the tests in Jetty 10 were carried over, but `@Disabled` in Jetty 12.0.x.

When the work on #8201 occurred (Multipart overhaul), _some_ of tests were un-disabled.
Other tests were un-disabled as time went on after that PR.

But the 500 vs 400 status code differences were not being tested in Jetty 10 vs Jetty 12 for the cases that matter, only a tiny fraction were being tested.
In Jetty 12 we were able to make fundamental changes and have tests pass as the tests were insufficient to catch the regression on many error cases.
This PR adds more tests for error cases and restores the behavior as seen in Jetty 10.

Comment (joakime) [reply jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/MultiPartFormInputStream.java]: @gregw the `total == 0` is the only case we are missing in the codebase in Jetty 12.
That would be the `testEmptyBodyMultipartForm()` test case showing that.

For Jetty 10, the following code catches this situation.

**Legacy parser**

https://github.com/jetty/jetty.project/blob/aa5d3d018c6fe242d1dd43d68b8ff43e808a2c41/jetty-server/src/main/java/org/eclipse/jetty/server/MultiPartInputStreamParser.java#L582-L583

**Spec compliant parser**

https://github.com/jetty/jetty.project/blob/aa5d3d018c6fe242d1dd43d68b8ff43e808a2c41/jetty-server/src/main/java/org/eclipse/jetty/server/MultiPartFormInputStream.java#L629-L632

Follow the links on those snippets to see the code around (before) those highlighted lines.

Comment (joakime) [reply jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/MultiPartFormInputStream.java]: @gregw the other cases, like `total >= 1`, or all whitespace, or other similar things are all currently handled by the codebases (both legacy and spec)

Comment (gregw) [reply jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/MultiPartFormInputStream.java]: @joakime I understand those cases are being handled by the code base.  What I'm asking is why can't the existing code be modified to also handle `total==0`?  why is a special test needed?  Isn't `total==0` just the edge case of the handle for all `total < boundary.length()`.

I've not look at the code in detail, but the  `if (parser.getState() != MultiPartParser.State.END)` looks like it should be able to handle `total==0`.  If no bytes are received than the parser will not be in END state, so an error can be generated by that existing code rather than writing a new test.

Comment (joakime) [reply jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/MultiPartFormInputStream.java]: I removed these lines and put the conversion of IOException to BadMessage in the Request.getParts() instead.

**Review by gregw — APPROVED** [review]
LGTM
